File: app/src/main/java/pk/aspirasoft/tasbih/data/CounterManager.java
package pk.aspirasoft.tasbih.data;

import android.app.Activity;

import java.util.ArrayList;

public class CounterManager extends ArrayList<Counter> {

    private static CounterManager ourInstance;
    private final Database database;

    private boolean isInitialized;

    private CounterManager(Activity activity) {
        database = Database.getInstance(activity);
        isInitialized = database.isInitialized();
        diskIn();
    }

    public static CounterManager getInstance(Activity activity) throws NullPointerException {
        if (ourInstance == null) {
            ourInstance = new CounterManager(activity);
        }

        return ourInstance;
    }

    public void addDefaultCounters(ArrayList<Counter> counters) {
        this.addAll(counters);
        diskOut();
        database.initialize();
    }

    public boolean isInitialized() {
        return isInitialized;
    }

    public void diskOut() {
        database.put(toString());
    }

    private void diskIn() {
        String counters = database.get();

        if (counters != null) {
            fromString(counters);
        }
    }

    public void fromString(String counters) {
        clear();

        for (String s : counters.split("\n")) {
            if (!s.replace("\t", "").replace(" ", "").equals("")) {
                try {
                    Counter counter = new Counter(s);
                    add(counter);
                } catch (ArrayIndexOutOfBoundsException ignored) {

                }
            }
        }
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        for (int i = 0; i < size(); i++) {
            string.append(get(i).toString()).append("\n");
        }
        string.append("");
        return string.toString();
    }

    @Override
    public boolean add(Counter object) {
        for (Counter c : this) {
            if (c.getName().equals(object.getName()))
                return false;
        }

        return super.add(object);
    }
}

File: android/app/src/main/java/pk/aspirasoft/tasbih/models/CounterManager.kt
package pk.aspirasoft.tasbih.models

import com.google.gson.Gson
import java.util.*

object CounterManager : ArrayList<Tasbih>() {

    init {
        diskIn()
    }

    override fun add(element: Tasbih): Boolean {
        return when {
            this.any { it.name.equals(element.name, true) } -> false
            super.add(element) -> {
                diskOut()
                true
            }
            else -> false
        }
    }

    override fun add(index: Int, element: Tasbih) {
        super.add(index, element)
        diskOut()
    }

    override fun addAll(elements: Collection<Tasbih>): Boolean {
        val r = super.addAll(elements)
        diskOut()
        return r
    }

    override fun addAll(index: Int, elements: Collection<Tasbih>): Boolean {
        val r = super.addAll(index, elements)
        diskOut()
        return r
    }

    override fun remove(element: Tasbih): Boolean {
        val r = super.remove(element)
        diskOut()
        return r
    }

    override fun removeAll(elements: Collection<Tasbih>): Boolean {
        val r = super.removeAll(elements)
        diskOut()
        return r
    }

    override fun removeAt(index: Int): Tasbih {
        val r = super.removeAt(index)
        diskOut()
        return r
    }

    override fun removeRange(fromIndex: Int, toIndex: Int) {
        super.removeRange(fromIndex, toIndex)
        diskOut()
    }

    override fun clear() {
        super.clear()
        diskOut()
    }

    fun replaceAll(elements: ArrayList<Tasbih>) {
        this.clear()
        this.addAll(elements)

        diskOut()
    }

    private fun diskIn() {
        Database.get("counters")?.let {
            fromJson(it)
        }
    }

    fun diskOut() {
        Database.put("counters", toString())
    }

    private fun fromJson(json: String) {
        Gson().fromJson(json, CounterManager::class.java)?.let {
            this.replaceAll(it)
        }
    }

    override fun toString(): String {
        return Gson().toJson(this)
    }

    override val size: Int
        get() {
            diskIn()
            return super.size
        }

}